contracts/MigrationContractStaging.cdc

```diff
@@ -0,0 +1,362 @@
+    /// to stage on its behalf), it should create a PRIVATE Host capability and publish it to the receiving account.
+    ///
+    access(all) resource Host {
+        /// Returns the resource owner's address
+        ///
+        access(all) view fun address(): Address {
+            return self.owner?.address ?? panic("Host is unowned!")
+        }
+    }
+
+    /********************
+            Capsule
+     ********************/
+
+    /// Resource that stores pending contract updates in a ContractUpdate struct. On staging a contract update for the
+    /// first time, a Capsule will be created and stored in this contract account. Any time a stageContract() call is
+    /// made again for the same contract, the code in the Capsule will be replaced. As you see, the Capsule is merely
+    /// intended to store the code, as contract updates will be executed by state migration across the network at the
+    /// Cadence 1.0 milestone.
+    ///
+    access(all) resource Capsule {
+        /// The address, name and code of the contract that will be updated.
+        access(self) let update: ContractUpdate
+
+        init(update: ContractUpdate) {
+            pre {
+                update.codeAsCadence() != nil: "Staged update code must be valid Cadence"
+                update.isValid(): "Target contract does not exist"
+            }
+            self.update = update
+        }
+
+        /// Returns the staged contract update in the form of a ContractUpdate struct.
+        ///
+        access(all) view fun getContractUpdate(): ContractUpdate {
+            return self.update
+        }
+
+        /// Replaces the staged contract code with the given hex-encoded Cadence code.
+        ///
+        access(contract) fun replaceCode(code: String) {
+            post {
+                self.update.codeAsCadence() != nil: "Staged update code must be valid Cadence"
+            }
+            self.update.replaceCode(code)
+            emit StagingStatusUpdated(
+                capsuleUUID: self.uuid,
+                address: self.update.address,
+                codeHash: code.decodeHex(),
+                contract: self.update.name,
+                action: "replace"
+            )
+        }
+    }
+
+    /********************
+            Admin
+     ********************/
+
+    /// Admin resource for updating the stagingCutoff value
+    ///
+    access(all) resource Admin {
+
+        /// Sets the block height at which updates can no longer be staged
+        ///
+        access(all) fun setStagingCutoff(atHeight: UInt64?) {
+            pre {
+                atHeight == nil || atHeight! > getCurrentBlock().height:
+                    "Height must be nil or greater than current block height"
+            }
+            emit StagingCutoffUpdated(old: MigrationContractStaging.stagingCutoff, new: atHeight)
+            MigrationContractStaging.stagingCutoff = atHeight
```

Comment:
Maybe use both argument label and parameter name, see https://www.swift.org/documentation/api-design-guidelines/#argument-labels:

```suggestion
        access(all) fun setStagingCutoff(at height: UInt64?) {
            pre {
                height == nil || height! > getCurrentBlock().height:
                    "Height must be nil or greater than current block height"
            }
            emit StagingCutoffUpdated(old: MigrationContractStaging.stagingCutoff, new: height)
            MigrationContractStaging.stagingCutoff = height
```